File: VB/PayrollBusiness.Business.vb
Imports PayrollBusiness.ServiceContracts
Imports PayrollDAL
Imports Framework.Data

Namespace PayrollBusiness.ServiceImplementations
    Partial Public Class PayrollBusiness
        Implements ServiceContracts.IPayrollBusiness

#Region "Calculate Salary"
        Public Function Load_Calculate_Load(ByVal lstEmp As List(Of PAEmployeeCalculateDTO), ByVal OrgId As Integer, ByVal PeriodId As Integer, ByVal IsDissolve As Integer, ByVal IsLoad As Integer, ByVal log As UserLog) As Boolean Implements ServiceContracts.IPayrollBusiness.Load_Calculate_Load
            Dim rep As New PayrollRepository
            Return rep.Load_Calculate_Load(lstEmp, OrgId, PeriodId, IsDissolve, IsLoad, log)
        End Function
        Public Function Calculate_data_sum(ByVal OrgId As Integer, ByVal PeriodId As Integer, ByVal IsDissolve As Integer, ByVal IsLoad As Integer, ByVal log As UserLog) As Boolean Implements ServiceContracts.IPayrollBusiness.Calculate_data_sum
            Dim rep As New PayrollRepository
            Return rep.Calculate_data_sum(OrgId, PeriodId, IsDissolve, IsLoad, log)
        End Function
        Public Function Load_data_sum(ByVal OrgId As Integer, ByVal PeriodId As Integer, ByVal IsDissolve As Integer, ByVal IsLoad As Integer, ByVal log As UserLog) As Boolean Implements ServiceContracts.IPayrollBusiness.Load_data_sum
            Dim rep As New PayrollRepository
            Return rep.Load_data_sum(OrgId, PeriodId, IsDissolve, IsLoad, log)
        End Function
        Public Function Calculate_data_temp(ByVal OrgId As Integer, ByVal PeriodId As Integer, ByVal IsDissolve As Integer, ByVal IsLoad As Integer, ByVal log As UserLog) As Boolean Implements ServiceContracts.IPayrollBusiness.Calculate_data_temp
            Dim rep As New PayrollRepository
            Return rep.Calculate_data_temp(OrgId, PeriodId, IsDissolve, IsLoad, log)
        End Function

        Public Function GetLitsCalculate(ByVal param As PA_ParamDTO, ByVal IsLoad As Integer,
                                          ByVal PageIndex As Integer, ByVal PageSize As Integer, ByVal PageType As Integer, ByRef TotalRow As Integer,
                                     ByVal log As UserLog, Optional ByVal Sorts As String = "CREATED_DATE DESC") As DataTable Implements ServiceContracts.IPayrollBusiness.GetLitsCalculate
            Dim rep As New PayrollRepository
            Return rep.GetLitsCalculate(param, IsLoad, PageIndex, PageSize, PageType, TotalRow, log)
        End Function
        Public Function GetListSalaryVisibleCol() As List(Of PAListSalariesDTO) Implements ServiceContracts.IPayrollBusiness.GetListSalaryVisibleCol
            Dim rep As New PayrollRepository
            Return rep.GetListSalaryVisibleCol()
        End Function

        Public Function ActiveOrDeactive(ByVal _param As PA_ParamDTO,
                                     ByVal log As UserLog) As Boolean Implements ServiceContracts.IPayrollBusiness.ActiveOrDeactive
            Dim rep As New PayrollRepository
            Return rep.ActiveOrDeactive(_param, log)
        End Function

#End Region

#Region "Import Salary"

        Public Function GetImportSalary(ByVal PeriodId As Integer, ByVal OrgId As Integer, ByVal IsDissolve As Integer, ByVal EmployeeId As String,
                                     ByVal PageIndex As Integer,
                                     ByVal PageSize As Integer,
                                     ByRef TotalRow As Integer,
                                     ByVal log As UserLog,
                                        Optional ByVal Sorts As String = "CREATED_DATE DESC") As DataTable Implements ServiceContracts.IPayrollBusiness.GetImportSalary
            Try
                Dim rep As New PayrollRepository
                Return rep.GetImportSalary(PeriodId, OrgId, IsDissolve, EmployeeId, PageIndex, PageSize, TotalRow, log, Sorts)
            Catch ex As Exception
                Throw ex
            End Try
        End Function
        Public Function GetSalaryList() As List(Of PAListSalariesDTO) Implements ServiceContracts.IPayrollBusiness.GetSalaryList
            Try
                Dim rep As New PayrollRepository
                Return rep.GetSalaryList()
            Catch ex As Exception
                Throw ex
            End Try
        End Function
        Public Function SaveImport(ByVal Period As Decimal, ByVal dtData As DataTable, ByVal lstColVal As List(Of String), ByVal log As UserLog, ByRef RecordSussces As Integer) As Boolean Implements ServiceContracts.IPayrollBusiness.SaveImport
            Try
                Dim rep As New PayrollRepository
                Return rep.SaveImport(Period, dtData, lstColVal, log, RecordSussces)
            Catch ex As Exception
                Throw ex
            End Try
        End Function
#End Region

#Region "IPORTAL - View phiếu lương"
        Public Function GetPayrollSheetSum(ByVal PeriodId As Integer, ByVal EmployeeId As String, ByVal log As UserLog,
                                        Optional ByVal Sorts As String = "CREATED_DATE DESC") As DataTable Implements ServiceContracts.IPayrollBusiness.GetPayrollSheetSum
            Try
                Dim rep As New PayrollRepository
                Return rep.GetPayrollSheetSum(PeriodId, EmployeeId, log, Sorts)
            Catch ex As Exception
                Throw ex
            End Try
        End Function


        Public Function CheckPeriod(ByVal PeriodId As Integer, ByVal EmployeeId As Decimal) As Boolean Implements ServiceContracts.IPayrollBusiness.CheckPeriod
            Try
                Dim rep As New PayrollRepository
                Return rep.CheckPeriod(PeriodId, EmployeeId)
            Catch ex As Exception
                Throw ex
            End Try
        End Function
#End Region

#Region "Báo cáo lương "
        Public Function GetReportById(ByVal _filter As Se_ReportDTO, ByVal PageIndex As Integer,
                                        ByVal PageSize As Integer,
                                        ByRef Total As Integer,
                                        log As UserLog,
                                        Optional ByVal Sorts As String = "CODE ASC") As List(Of Se_ReportDTO) Implements ServiceContracts.IPayrollBusiness.GetReportById
            Using rep As New PayrollRepository
                Try

                    Dim lst = rep.GetReportById(_filter, PageIndex, PageSize, Total, log, Sorts)
                    Return lst
                Catch ex As Exception
                    Throw ex
                End Try
            End Using
        End Function

        Public Function ExportReport(ByVal sPkgName As String,
                                     ByVal sMoth As Decimal,
                                     ByVal sYear As Decimal,
                                     ByVal sOrgId As String,
                                     ByVal IsDissolve As Decimal,
                                     ByVal log As UserLog) As DataSet Implements ServiceContracts.IPayrollBusiness.ExportReport
            Using rep As New PayrollRepository
                Try
                    Dim lst = rep.ExportReport(sPkgName, sMoth, sYear, sOrgId, IsDissolve, log)
                    Return lst
                Catch ex As Exception
                    Throw ex
                End Try
            End Using
        End Function

        Public Function ExportReport_005(ByVal sPkgName As String,
                                         ByVal sMoth As Decimal,
                                         ByVal sYear As Decimal,
                                     ByVal sOrgId As String,
                                         ByVal IsDissolve As Decimal,
                                         ByVal log As UserLog) As DataSet _
                                     Implements ServiceContracts.IPayrollBusiness.ExportReport_005
            Using rep As New PayrollRepository
                Try
                    Dim lst = rep.ExportReport_005(sPkgName, sMoth, sYear, sOrgId, IsDissolve, log)
                    Return lst
                Catch ex As Exception
                    Throw ex
                End Try
            End Using
        End Function

        Public Function ExportReport_008(ByVal sPkgName As String,
                                         ByVal sMoth As Decimal,
                                         ByVal sYear As Decimal,
                                     ByVal sOrgId As String,
                                         ByVal IsDissolve As Decimal,
                                         ByVal log As UserLog) As DataSet _
                                     Implements ServiceContracts.IPayrollBusiness.ExportReport_008
            Using rep As New PayrollRepository
                Try
                    Dim lst = rep.ExportReport_008(sPkgName, sMoth, sYear, sOrgId, IsDissolve, log)
                    Return lst
                Catch ex As Exception
                    Throw ex
                End Try
            End Using
        End Function

        Public Function ExportReport_010(ByVal sPkgName As String,
                                         ByVal sMoth As Decimal,
                                         ByVal sYear As Decimal,
                                     ByVal sOrgId As String,
                                         ByVal IsDissolve As Decimal,
                                         ByVal log As UserLog) As DataSet _
                                     Implements ServiceContracts.IPayrollBusiness.ExportReport_010
            Using rep As New PayrollRepository
                Try
                    Dim lst = rep.ExportReport_010(sPkgName, sMoth, sYear, sOrgId, IsDissolve, log)
                    Return lst
                Catch ex As Exception
                    Throw ex
                End Try
            End Using
        End Function

        Public Function ExportReport_014(ByVal sPkgName As String,
                                         ByVal sMoth As Decimal,
                                         ByVal sYear As Decimal,
                                     ByVal sOrgId As String,
                                         ByVal IsDissolve As Decimal,
                                         ByVal log As UserLog) As DataSet _
                                     Implements ServiceContracts.IPayrollBusiness.ExportReport_014
            Using rep As New PayrollRepository
                Try
                    Dim lst = rep.ExportReport_014(sPkgName, sMoth, sYear, sOrgId, IsDissolve, log)
                    Return lst
                Catch ex As Exception
                    Throw ex
                End Try
            End Using
        End Function

        Public Function ExportPhieuLuong(ByVal lstEmployee As List(Of Decimal),
                                     ByVal orgID As Decimal?,
                                     ByVal isDissolve As Decimal?,
                                     ByVal periodID As Decimal,
                                  ByVal log As UserLog) As DataTable Implements ServiceContracts.IPayrollBusiness.ExportPhieuLuong
            Using rep As New PayrollRepository
                Try
                    Dim lst = rep.ExportPhieuLuong(lstEmployee, orgID, isDissolve, periodID, log)
                    Return lst
                Catch ex As Exception
                    Throw ex
                End Try
            End Using
        End Function

        Public Function ActionSendPayslip(ByVal lstEmployee As List(Of Decimal),
                                     ByVal orgID As Decimal?,
                                     ByVal isDissolve As Decimal?,
                                     ByVal periodID As Decimal,
                                     ByVal log As UserLog) As Boolean Implements ServiceContracts.IPayrollBusiness.ActionSendPayslip
            Using rep As New PayrollRepository
                Try
                    Return rep.ActionSendPayslip(lstEmployee, orgID, isDissolve, periodID, log)
                Catch ex As Exception
                    Throw ex
                End Try
            End Using
        End Function

#End Region

#Region "LOG"
        Function GetActionLog(ByVal _filter As PA_ACTION_LOGDTO,
                                 Optional ByRef Total As Integer = 0,
                                 Optional ByVal PageIndex As Integer = 0,
                                 Optional ByVal PageSize As Integer = Integer.MaxValue,
                                 Optional ByVal Sorts As String = "ACTION_DATE desc") As List(Of PA_ACTION_LOGDTO) Implements IPayrollBusiness.GetActionLog

            Using rep As New PayrollRepository
                Try
                    Dim lst = rep.GetActionLog(_filter, Total, PageIndex, PageSize, Sorts)
                    Return lst
                Catch ex As Exception
                    Throw ex
                End Try
            End Using
        End Function

        Function DeleteActionLogsPA(ByVal lstDeleteIds As List(Of Decimal)) As Integer Implements IPayrollBusiness.DeleteActionLogsPA

            Using rep As New PayrollRepository
                Try
                    Dim lst = rep.DeleteActionLogsPA(lstDeleteIds)
                    Return lst
                Catch ex As Exception
                    Throw ex
                End Try
            End Using
        End Function
#End Region

#Region "Seniority"

        Function CalSeniorityProcess(ByVal _filter As PASeniorityProcessDTO, ByVal log As UserLog) As Boolean _
                                    Implements IPayrollBusiness.CalSeniorityProcess

            Using rep As New PayrollRepository
                Try
                    Dim lst = rep.CalSeniorityProcess(_filter, log)
                    Return lst
                Catch ex As Exception
                    Throw ex
                End Try
            End Using
        End Function

        Function GetSeniorityProcess(ByVal _filter As PASeniorityProcessDTO,
                                        ByVal PageIndex As Integer,
                                        ByVal PageSize As Integer,
                                        ByRef Total As Integer, ByVal log As UserLog,
                                        Optional ByVal Sorts As String = "EMPLOYEE_CODE") As List(Of PASeniorityProcessDTO) _
                                    Implements IPayrollBusiness.GetSeniorityProcess

            Using rep As New PayrollRepository
                Try
                    Dim lst = rep.GetSeniorityProcess(_filter, PageIndex, PageSize, Total, log, Sorts)
                    Return lst
                Catch ex As Exception
                    Throw ex
                End Try
            End Using
        End Function

        Function GetSeniorityProcessImport(ByVal _filter As PASeniorityProcessDTO,
                                        ByVal log As UserLog) As DataTable _
                                    Implements IPayrollBusiness.GetSeniorityProcessImport

            Using rep As New PayrollRepository
                Try
                    Dim lst = rep.GetSeniorityProcessImport(_filter, log)
                    Return lst
                Catch ex As Exception
                    Throw ex
                End Try
            End Using
        End Function

        Function SaveSeniorityProcessImport(ByVal dtData As DataTable,
                                               ByVal periodId As Decimal,
                                               ByVal log As UserLog) _
                                    Implements IPayrollBusiness.SaveSeniorityProcessImport

            Using rep As New PayrollRepository
                Try
                    Dim lst = rep.SaveSeniorityProcessImport(dtData, periodId, log)
                    Return lst
                Catch ex As Exception
                    Throw ex
                End Try
            End Using
        End Function

        Function CalSeniorityMonthly(ByVal _filter As PASeniorityMonthlyDTO, ByVal log As UserLog) As Boolean _
                                    Implements IPayrollBusiness.CalSeniorityMonthly

            Using rep As New PayrollRepository
                Try
                    Dim lst = rep.CalSeniorityMonthly(_filter, log)
                    Return lst
                Catch ex As Exception
                    Throw ex
                End Try
            End Using
        End Function


        Function GetSeniorityMonthly(ByVal _filter As PASeniorityMonthlyDTO,
                                        ByVal PageIndex As Integer,
                                        ByVal PageSize As Integer,
                                        ByRef Total As Integer, ByVal log As UserLog,
                                        Optional ByVal Sorts As String = "EMPLOYEE_CODE") As List(Of PASeniorityMonthlyDTO) _
                                    Implements IPayrollBusiness.GetSeniorityMonthly

            Using rep As New PayrollRepository
                Try
                    Dim lst = rep.GetSeniorityMonthly(_filter, PageIndex, PageSize, Total, log, Sorts)
                    Return lst
                Catch ex As Exception
                    Throw ex
                End Try
            End Using
        End Function

#End Region


#Region "MailRemark"

        Public Function GetMailRemark(ByVal _filter As PAMailRemarkDTO,
                                   ByVal PageIndex As Integer,
                                        ByVal PageSize As Integer,
                                        ByRef Total As Integer, ByVal log As UserLog,
                                        Optional ByVal Sorts As String = "CREATED_DATE desc") As List(Of PAMailRemarkDTO) Implements ServiceContracts.IPayrollBusiness.GetMailRemark
            Try
                Dim rep As New PayrollRepository
                Return rep.GetMailRemark(_filter, PageIndex, PageSize, Total, log, Sorts)
            Catch ex As Exception

                Throw ex
            End Try
        End Function

        Public Function InsertMailRemark(ByVal objMailRemark As PAMailRemarkDTO, ByVal log As UserLog, ByRef gID As Decimal) As Boolean Implements ServiceContracts.IPayrollBusiness.InsertMailRemark
            Try
                Dim rep As New PayrollRepository
                Return rep.InsertMailRemark(objMailRemark, log, gID)
            Catch ex As Exception

                Throw ex
            End Try
        End Function

        Public Function ModifyMailRemark(ByVal objMailRemark As PAMailRemarkDTO, ByVal log As UserLog, ByRef gID As Decimal) As Boolean Implements ServiceContracts.IPayrollBusiness.ModifyMailRemark
            Try
                Dim rep As New PayrollRepository
                Return rep.ModifyMailRemark(objMailRemark, log, gID)
            Catch ex As Exception
                Throw ex
            End Try
        End Function

        Public Function DeleteMailRemark(ByVal lstID As List(Of Decimal)) As Boolean Implements ServiceContracts.IPayrollBusiness.DeleteMailRemark
            Try
                Dim rep As New PayrollRepository
                Return rep.DeleteMailRemark(lstID)
            Catch ex As Exception
                Throw ex
            End Try
        End Function

#End Region


    End Class
End Namespace

File: C#/PayrollBusiness.Business.cs
using System;
using PayrollBusiness.ServiceContracts;
using PayrollDAL;
using Framework.Data;

namespace PayrollBusiness.ServiceImplementations
{
    public partial class PayrollBusiness : ServiceContracts.IPayrollBusiness
    {
        #region "Calculate Salary"
        public bool Load_Calculate_Load(List<PAEmployeeCalculateDTO> lstEmp, int OrgId, int PeriodId, int IsDissolve, int IsLoad, UserLog log)
        {
            PayrollRepository rep = new PayrollRepository();
            return rep.Load_Calculate_Load(lstEmp, OrgId, PeriodId, IsDissolve, IsLoad, log);
        }
        public bool Calculate_data_sum(int OrgId, int PeriodId, int IsDissolve, int IsLoad, UserLog log)
        {
            PayrollRepository rep = new PayrollRepository();
            return rep.Calculate_data_sum(OrgId, PeriodId, IsDissolve, IsLoad, log);
        }
        public bool Load_data_sum(int OrgId, int PeriodId, int IsDissolve, int IsLoad, UserLog log)
        {
            PayrollRepository rep = new PayrollRepository();
            return rep.Load_data_sum(OrgId, PeriodId, IsDissolve, IsLoad, log);
        }
        public bool Calculate_data_temp(int OrgId, int PeriodId, int IsDissolve, int IsLoad, UserLog log)
        {
            PayrollRepository rep = new PayrollRepository();
            return rep.Calculate_data_temp(OrgId, PeriodId, IsDissolve, IsLoad, log);
        }

        public DataTable GetLitsCalculate(PA_ParamDTO param, int IsLoad, int PageIndex, int PageSize, int PageType, ref int TotalRow, UserLog log, string Sorts = "CREATED_DATE DESC")
        {
            PayrollRepository rep = new PayrollRepository();
            return rep.GetLitsCalculate(param, IsLoad, PageIndex, PageSize, PageType, TotalRow, log);
        }
        public List<PAListSalariesDTO> GetListSalaryVisibleCol()
        {
            PayrollRepository rep = new PayrollRepository();
            return rep.GetListSalaryVisibleCol();
        }

        public bool ActiveOrDeactive(PA_ParamDTO _param, UserLog log)
        {
            PayrollRepository rep = new PayrollRepository();
            return rep.ActiveOrDeactive(_param, log);
        }
        #endregion

        #region "Import Salary"
        public DataTable GetImportSalary(int PeriodId, int OrgId, int IsDissolve, string EmployeeId, int PageIndex, int PageSize, ref int TotalRow, UserLog log, string Sorts = "CREATED_DATE DESC")
        {
            try
            {
                PayrollRepository rep = new PayrollRepository();
                return rep.GetImportSalary(PeriodId, OrgId, IsDissolve, EmployeeId, PageIndex, PageSize, TotalRow, log, Sorts);
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }
        public List<PAListSalariesDTO> GetSalaryList()
        {
            try
            {
                PayrollRepository rep = new PayrollRepository();
                return rep.GetSalaryList();
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }
        public bool SaveImport(decimal Period, DataTable dtData, List<string> lstColVal, UserLog log, ref int RecordSussces)
        {
            try
            {
                PayrollRepository rep = new PayrollRepository();
                return rep.SaveImport(Period, dtData, lstColVal, log, RecordSussces);
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }
        #endregion

        #region "IPORTAL - View phiếu lương"
        public DataTable GetPayrollSheetSum(int PeriodId, string EmployeeId, UserLog log, string Sorts = "CREATED_DATE DESC")
        {
            try
            {
                PayrollRepository rep = new PayrollRepository();
                return rep.GetPayrollSheetSum(PeriodId, EmployeeId, log, Sorts);
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }


        public bool CheckPeriod(int PeriodId, decimal EmployeeId)
        {
            try
            {
                PayrollRepository rep = new PayrollRepository();
                return rep.CheckPeriod(PeriodId, EmployeeId);
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }
        #endregion

        #region "Báo cáo lương "
        public List<Se_ReportDTO> GetReportById(Se_ReportDTO _filter, int PageIndex, int PageSize, ref int Total, UserLog log, string Sorts = "CODE ASC")
        {
            using (PayrollRepository rep = new PayrollRepository())
            {
                try
                {
                    var lst = rep.GetReportById(_filter, PageIndex, PageSize, Total, log, Sorts);
                    return lst;
                }
                catch (Exception ex)
                {
                    throw ex;
                }
            }
        }

        public DataSet ExportReport(string sPkgName, decimal sMoth, decimal sYear, string sOrgId, decimal IsDissolve, UserLog log)
        {
            using (PayrollRepository rep = new PayrollRepository())
            {
                try
                {
                    var lst = rep.ExportReport(sPkgName, sMoth, sYear, sOrgId, IsDissolve, log);
                    return lst;
                }
                catch (Exception ex)
                {
                    throw ex;
                }
            }
        }

        public DataSet ExportReport_005(string sPkgName, decimal sMoth, decimal sYear, string sOrgId, decimal IsDissolve, UserLog log)
        {
            using (PayrollRepository rep = new PayrollRepository())
            {
                try
                {
                    var lst = rep.ExportReport_005(sPkgName, sMoth, sYear, sOrgId, IsDissolve, log);
                    return lst;
                }
                catch (Exception ex)
                {
                    throw ex;
                }
            }
        }

        public DataSet ExportReport_008(string sPkgName, decimal sMoth, decimal sYear, string sOrgId, decimal IsDissolve, UserLog log)
        {
            using (PayrollRepository rep = new PayrollRepository())
            {
                try
                {
                    var lst = rep.ExportReport_008(sPkgName, sMoth, sYear, sOrgId, IsDissolve, log);
                    return lst;
                }
                catch (Exception ex)
                {
                    throw ex;
                }
            }
        }

        public DataSet ExportReport_010(string sPkgName, decimal sMoth, decimal sYear, string sOrgId, decimal IsDissolve, UserLog log)
        {
            using (PayrollRepository rep = new PayrollRepository())
            {
                try
                {
                    var lst = rep.ExportReport_010(sPkgName, sMoth, sYear, sOrgId, IsDissolve, log);
                    return lst;
                }
                catch (Exception ex)
                {
                    throw ex;
                }
            }
        }

        public DataSet ExportReport_014(string sPkgName, decimal sMoth, decimal sYear, string sOrgId, decimal IsDissolve, UserLog log)
        {
            using (PayrollRepository rep = new PayrollRepository())
            {
                try
                {
                    var lst = rep.ExportReport_014(sPkgName, sMoth, sYear, sOrgId, IsDissolve, log);
                    return lst;
                }
                catch (Exception ex)
                {
                    throw ex;
                }
            }
        }

        public DataTable ExportPhieuLuong(List<decimal> lstEmployee, decimal? orgID, decimal? isDissolve, decimal periodID, UserLog log)
        {
            using (PayrollRepository rep = new PayrollRepository())
            {
                try
                {
                    var lst = rep.ExportPhieuLuong(lstEmployee, orgID, isDissolve, periodID, log);
                    return lst;
                }
                catch (Exception ex)
                {
                    throw ex;
                }
            }
        }

        public bool ActionSendPayslip(List<decimal> lstEmployee, decimal? orgID, decimal? isDissolve, decimal periodID, UserLog log)
        {
            using (PayrollRepository rep = new PayrollRepository())
            {
                try
                {
                    return rep.ActionSendPayslip(lstEmployee, orgID, isDissolve, periodID, log);
                }
                catch (Exception ex)
                {
                    throw ex;
                }
            }
        }
        #endregion

        #region "LOG"
        public List<PA_ACTION_LOGDTO> GetActionLog(PA_ACTION_LOGDTO _filter, ref int Total = 0, int PageIndex = 0, int PageSize = int.MaxValue, string Sorts = "ACTION_DATE desc")
        {
            using (PayrollRepository rep = new PayrollRepository())
            {
                try
                {
                    var lst = rep.GetActionLog(_filter, Total, PageIndex, PageSize, Sorts);
                    return lst;
                }
                catch (Exception ex)
                {
                    throw ex;
                }
            }
        }

        public int DeleteActionLogsPA(List<decimal> lstDeleteIds)
        {
            using (PayrollRepository rep = new PayrollRepository())
            {
                try
                {
                    var lst = rep.DeleteActionLogsPA(lstDeleteIds);
                    return lst;
                }
                catch (Exception ex)
                {
                    throw ex;
                }
            }
        }
        #endregion

        #region "Seniority"
        public bool CalSeniorityProcess(PASeniorityProcessDTO _filter, UserLog log)
        {
            using (PayrollRepository rep = new PayrollRepository())
            {
                try
                {
                    var lst = rep.CalSeniorityProcess(_filter, log);
                    return lst;
                }
                catch (Exception ex)
                {
                    throw ex;
                }
            }
        }

        public List<PASeniorityProcessDTO> GetSeniorityProcess(PASeniorityProcessDTO _filter, int PageIndex, int PageSize, ref int Total, UserLog log, string Sorts = "EMPLOYEE_CODE")
        {
            using (PayrollRepository rep = new PayrollRepository())
            {
                try
                {
                    var lst = rep.GetSeniorityProcess(_filter, PageIndex, PageSize, Total, log, Sorts);
                    return lst;
                }
                catch (Exception ex)
                {
                    throw ex;
                }
            }
        }

        public DataTable GetSeniorityProcessImport(PASeniorityProcessDTO _filter, UserLog log)
        {
            using (PayrollRepository rep = new PayrollRepository())
            {
                try
                {
                    var lst = rep.GetSeniorityProcessImport(_filter, log);
                    return lst;
                }
                catch (Exception ex)
                {
                    throw ex;
                }
            }
        }

        public void SaveSeniorityProcessImport(DataTable dtData, decimal periodId, UserLog log)
        {
            using (PayrollRepository rep = new PayrollRepository())
            {
                try
                {
                    var lst = rep.SaveSeniorityProcessImport(dtData, periodId, log);
                    return lst;
                }
                catch (Exception ex)
                {
                    throw ex;
                }
            }
        }

        public bool CalSeniorityMonthly(PASeniorityMonthlyDTO _filter, UserLog log)
        {
            using (PayrollRepository rep = new PayrollRepository())
            {
                try
                {
                    var lst = rep.CalSeniorityMonthly(_filter, log);
                    return lst;
                }
                catch (Exception ex)
                {
                    throw ex;
                }
            }
        }


        public List<PASeniorityMonthlyDTO> GetSeniorityMonthly(PASeniorityMonthlyDTO _filter, int PageIndex, int PageSize, ref int Total, UserLog log, string Sorts = "EMPLOYEE_CODE")
        {
            using (PayrollRepository rep = new PayrollRepository())
            {
                try
                {
                    var lst = rep.GetSeniorityMonthly(_filter, PageIndex, PageSize, Total, log, Sorts);
                    return lst;
                }
                catch (Exception ex)
                {
                    throw ex;
                }
            }
        }
        #endregion



        #region "MailRemark"
        public List<PAMailRemarkDTO> GetMailRemark(PAMailRemarkDTO _filter, int PageIndex, int PageSize, ref int Total, UserLog log, string Sorts = "CREATED_DATE desc")
        {
            try
            {
                PayrollRepository rep = new PayrollRepository();
                return rep.GetMailRemark(_filter, PageIndex, PageSize, Total, log, Sorts);
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        public bool InsertMailRemark(PAMailRemarkDTO objMailRemark, UserLog log, ref decimal gID)
        {
            try
            {
                PayrollRepository rep = new PayrollRepository();
                return rep.InsertMailRemark(objMailRemark, log, gID);
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        public bool ModifyMailRemark(PAMailRemarkDTO objMailRemark, UserLog log, ref decimal gID)
        {
            try
            {
                PayrollRepository rep = new PayrollRepository();
                return rep.ModifyMailRemark(objMailRemark, log, gID);
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        public bool DeleteMailRemark(List<decimal> lstID)
        {
            try
            {
                PayrollRepository rep = new PayrollRepository();
                return rep.DeleteMailRemark(lstID);
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }
        #endregion
    }
}
